kv/kvserver: TestTerm failed

kv/kvserver.TestTerm [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4480635&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4480635&tab=artifacts#/) on master @ [bf13dc6061ab080c8c4306ebabb5065331877508](https://github.com/cockroachdb/cockroach/commits/bf13dc6061ab080c8c4306ebabb5065331877508):


```
=== RUN   TestTerm
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestTerm391782973
    test_log_scope.go:80: use -show-logs to present logs inline
=== CONT  TestTerm
    replica_test.go:7424: -- test log scope end --
--- FAIL: TestTerm (0.13s)
=== RUN   TestTerm/loosely-coupled=true
    replica_test.go:7386: expected ErrCompacted, got %!s(<nil>)
    --- FAIL: TestTerm/loosely-coupled=true (0.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTerm.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13461

--- a/pkg/kv/kvserver/raft_log_queue_test.go
+++ b/pkg/kv/kvserver/raft_log_queue_test.go
@@ -925,12 +925,19 @@ func waitForTruncationForTesting(
 			// Flush the engine to advance durability, which triggers truncation.
 			require.NoError(t, r.Engine().Flush())
 		}
-		// FirstIndex has changed.
+		// FirstIndex should have changed.
 		firstIndex, err := r.GetFirstIndex()
 		require.NoError(t, err)
 		if firstIndex != newFirstIndex {
 			return errors.Errorf("expected firstIndex == %d, got %d", newFirstIndex, firstIndex)
 		}
+		// Some low-level tests also look at the raftEntryCache or sideloaded
+		// storage, which are updated after, and non-atomically with the change to
+		// first index (latter holds Replica.mu). Since the raftLogTruncator holds Replica.raftMu
+		// for the duration of its work, we can, by acquiring and releasing raftMu here, ensure
+		// that we have waited for it to finish.
+		r.raftMu.Lock()
+		defer r.raftMu.Unlock()
 		return nil
 	})
 }

